anonymizer/OpenDiffix.Service/Led.fs

```diff
@@ -0,0 +1,176 @@
+      CountArguments = Collections.Generic.List<Value list>()
+      ExecutionContext =
+        { executionContext with
+            NoiseLayers = { executionContext.NoiseLayers with BucketSeed = bucketSeed }
+        }
+      // Aggregator values are computed later
+      OutputRow = Array.init rowSize (fun i -> if i < groupingLabelsLength then group.[i] else Null)
+      IsLowCount = false
+    }
+
+  let state = Collections.Generic.Dictionary<Row, BucketState>(Row.equalityComparer)
+
+  // Group rows into buckets and keep track of low-count/count arguments
+  for row in Executor.execute executionContext childPlan do
+    let group = groupingLabels |> Array.map (Expression.evaluate row)
+    let argEvaluator = Expression.evaluate row
+
+    let bucket =
+      match state.TryGetValue(group) with
+      | true, bucketState -> bucketState
+      | false, _ ->
+        let bucketState = makeBucket group
+        state.[group] <- bucketState
+        bucketState
+
+    bucket.Aggregators
+    |> Array.iteri (fun i aggregator ->
+      let args = aggArgs.[i] |> List.map argEvaluator
+
+      if i = lowCountIndex then bucket.LowCountArguments.Add(args)
+      else if i = countIndex then bucket.CountArguments.Add(args)
+
+      args |> aggregator.Transition
+    )
+
+  // Compute aggregator final values
+  let buckets =
+    state
+    |> Seq.map (fun pair ->
+      let bucket = pair.Value
+      let executionContext = bucket.ExecutionContext
+      let outputRow = bucket.OutputRow
+
+      bucket.Aggregators
+      |> Array.iteri (fun i aggregator ->
+        if i = lowCountIndex then
+          let isLowCountValue = aggregator.Final executionContext
+          bucket.IsLowCount <- Value.unwrapBoolean isLowCountValue
+          outputRow.[groupingLabelsLength + i] <- isLowCountValue
+        else
+          outputRow.[groupingLabelsLength + i] <- aggregator.Final executionContext
+      )
+
+      bucket
+    )
+    |> Seq.toArray
+
+  // Merge victim buckets into sibling buckets
+  buckets
+  |> Array.iteri (fun i victimBucket ->
+    if victimBucket.IsLowCount then
+      let victimRow = victimBucket.OutputRow
+      let siblingsPerColumn = Array.init groupingLabelsLength (fun _i -> Collections.Generic.List<BucketState>(2))
+
+      buckets
+      |> Array.iteri (fun j otherBucket ->
+        if i <> j then
+          let nonMatchingColumn = findSingleNonMatchingColumn groupingLabelsLength victimRow otherBucket.OutputRow
+
+          if nonMatchingColumn > -1 then
+            let listOfSiblings = siblingsPerColumn.[nonMatchingColumn]
+            // We don't need to store all, just whether there are 0, 1, or more siblings
+            if listOfSiblings.Count < 2 then listOfSiblings.Add(otherBucket)
+      )
+
+      let hasUnknownColumn = siblingsPerColumn |> Array.exists (fun list -> list.Count = 0)
+
+      if hasUnknownColumn then
+        siblingsPerColumn
+        |> Array.iter (fun list ->
```

Comment:
A `for` loop would work well here.

Reply:
With the pattern match `iter` works better now.